packages/flutter/lib/src/material/text_selection_toolbar.dart

```diff
@@ -0,0 +1,742 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart' show listEquals;
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+
+import 'debug.dart';
+import 'icon_button.dart';
+import 'icons.dart';
+import 'material.dart';
+import 'material_localizations.dart';
+
+// Minimal padding from all edges of the selection toolbar to all edges of the
+// viewport.
+const double _kToolbarScreenPadding = 8.0;
+const double _kToolbarHeight = 44.0;
+
+// The type for a Function that builds a toolbar's container with the given
+// child.
+typedef _ToolbarBuilder = Widget Function(BuildContext context, Widget child);
+
+/// A fully-functional Material-style text selection toolbar.
+///
+/// Tries to position itself above [anchorAbove], but if it doesn't fit, then
+/// it positions itself below [anchorBelow].
+///
+/// If any children don't fit in the menu, an overflow menu will automatically
+/// be created.
+class TextSelectionToolbar extends StatelessWidget {
+  /// Creates an instance of TextSelectionToolbar.
+  const TextSelectionToolbar({
+    Key? key,
+    required this.anchorAbove,
+    required this.anchorBelow,
+    this.toolbarBuilder = _defaultToolbarBuilder,
+    required this.children,
+  }) : super(key: key);
+
+  /// The focal point above which the toolbar attempts to position itself.
+  ///
+  /// If there is not enough room above before reaching the top of the screen,
+  /// then the toolbar will position itself below [anchorBelow].
+  final Offset anchorAbove;
+
+  /// The focal point below which the toolbar attempts to position itself, if it
+  /// doesn't fit above [anchorAbove].
+  final Offset anchorBelow;
+
+  /// The children that will be displayed in the text selection toolbar.
+  ///
+  /// Typically these are buttons.
+  ///
+  /// See also:
+  ///   * [TextSelectionToolbarTextButton], which builds a default Material-
+  ///     style text selection toolbar text button.
+  final List<Widget> children;
+
+  /// Builds the toolbar container.
+  ///
+  /// Useful for customizing the high-level background of the toolbar. The given
+  /// child Widget will contain all of the [children].
+  final _ToolbarBuilder toolbarBuilder;
```

Comment:
I guess we should, since users will be passing in their own toolbarBuilders.  I thought about making this more generic and maybe naming it ChildWidgetBuilder, but since it's possible that we'll need to add a new toolbar-specific parameter to this in the future, I'll keep the name as ToolbarBuilder.